File: 3/2.js
const fs = require('fs');
const { get_val } = require('./Funcs.js');

(function () {
  const contents = fs.readFileSync("input.txt").toString();
  let lines = contents.split("\r\n");
  const groups = [[]];
  lines.forEach(line => {
    if (groups[groups.length - 1].length === 3) {
      groups.push([line]);
    } else {
      groups[groups.length - 1].push(line);
    }
  });
  const values = groups.filter(group => group.length === 3).map(group => {
    const g0 = group[0].split("");
    const g1 = group[1].split("");
    const g2 = group[2].split("");
    const int = g0.filter(x => g1.includes(x) && g2.includes(x));
    return get_val(int[0]);
  });
  const sum = values.reduce((a, b) => a + b, 0);
  console.log(sum);
})();

File: 3/2.hs
import System.IO
import Data.List.Split
import Funcs

-- Split a list into groups
split_groups :: Int -> [a] -> [[a]]
split_groups _ [] = []
split_groups n lst = (take n lst):(split_groups n (drop n lst))

main = do
  h <- openFile "input.txt" ReadMode
  contents <- hGetContents h
  let
    dups =
      map (\[a,b,c] -> head $ rem_dups $ intersect c $ intersect a b) $ -- Get letters occuring in both lists. head of string returns char.
      map (\t -> map (\s -> tail $ splitOn "" s) t) $ -- Split text into lists, remove first element ("")
      split_groups 3 $ -- Split into groups of 3
      lines contents -- Get lines (rounds)
    vals = map (get_val.head) dups -- Get value of each duplicate
  print $ sum vals
  hClose h